### PR TITLE
feat: add radom checkout session integration for new orders

### PR DESCRIPTION
**Pull request by clD11**
### Summary

This PR adds Radom checkout session integration for new orders.

Security Review: https://github.com/brave/reviews/issues/1719

resolves https://github.com/brave-intl/bat-go/issues/2633

### Type of Change

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before Requesting Review

- [x] Does your code build cleanly without any errors or warnings?
- [x] Have you used auto closing keywords?
- [x] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [x] Have you used intuitive function, variable and other naming?
- [x] Have you requested security and/or privacy review if needed
- [x] Have you performed a self review of this PR?

### CODE REVIEW

**Review by kdenhartog — APPROVED** [review]
Confirmed this is disabled by default at the moment. This LGTM from a privacy perspective.

https://github.com/brave-intl/bat-go/blob/99437001d3ddf523f16073ecb255af0bfb898273/services/grant/cmd/grant.go#L146

https://github.com/brave-intl/grant-ops/blob/master/ecs-vars-prod.env#L59

https://github.com/brave-intl/bat-go/pull/2627/files#diff-225dad392327a49c7aa0c5287edf78e66ae94759a42c08d4207299c0e0d6e895R235

File: services/skus/service_test.go [hunk 1]
```diff
@@ -8,14 +8,14 @@ import (
 	"testing"
 	"time"
 
+	"github.com/brave-intl/bat-go/libs/datastore"
```
Comment (pavelbrm) [inline services/skus/service_test.go:11]: issue: Something is up with the settings of your IDE. This is not the sorting we want, please revert the change.

File: services/skus/service.go [hunk 2]
```diff
@@ -2488,3 +2547,42 @@ func shouldUpdateOrderStripeSubID(ord *model.Order, subID string) bool {
 
 	return false
 }
+
+func newRadomGateway(env string) (radom.Gateway, error) {
+	switch env {
+	case "development", "staging":
+		return radom.Gateway{
+			Managed: radom.Managed{
+				Methods: []radom.Method{
```
Comment (pavelbrm) [inline services/skus/service.go:2556]: suggestion: Since the slice is going to escape to the heap anyway (since it's returned as part of a copy), we might as well return just a pointer to `radom.Gateway` and use it as pointer as opposed to allowing copies.

File: services/skus/service_nonint_test.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ import (
 	"time"
 
 	"github.com/awa/go-iap/appstore"
+	"github.com/brave-intl/bat-go/services/skus/radom"
```
Comment (pavelbrm) [inline services/skus/service_nonint_test.go:14]: nit: This import should go to the last group, this is a place for 3rd-party imports.

File: services/skus/service_nonint_test.go [hunk 4]
```diff
@@ -2695,6 +2696,365 @@ func TestShouldUpdateOrderStripeSubID(t *testing.T) {
...
+				req: &model.CreateOrderRequestNew{
+					RadomMetadata: &model.OrderRadomMetadata{
+						SuccessURI: "https://invalid%.com",
+					},
+				},
+			},
+			exp: tcExpected{
+				mustErr: func(t must.TestingT, err error, i ...interface{}) {
+					must.NotNil(t, err)
+				},
```
Comment (pavelbrm) [inline services/skus/service_nonint_test.go:2731]: issue: This way of expressing the error has two issues: 1) it couples expectations with a specific library which [this way of expression](https://github.com/brave-intl/bat-go/blob/master/services/skus/service_nonint_test.go#L3899-L4063) does not do; and 2) it makes the expectations harder to read and comprehend, compared to [the same way](https://github.com/brave-intl/bat-go/blob/master/services/skus/service_nonint_test.go#L3899-L4063), where the logic of dealing with errors is expressed in the test code itself.

The given example (and that's how the majority of tests is and has been written now) shifts focus to the actual expectation (i.e. error), and deals with the handling as part of the testing code, only once. The proposed way does the opposite, which has a negative effect – it makes the expectation declarations noisy at the expense of simplifying the testing code by one line at most. I would much rather have simple declarations and deal with the consequences in the testing code. The examples are plentiful now across the entire SKUs service, and there has never been a case _yet_ where it would not work. I would bet if it does not, then the code that is being tested is not well expressed enough.

**Review by pavelbrm — COMMENTED** [review]
Overall looking good. Not ready to approve, and would like to see most of the comments addressed before this gets merged.

**Review by pavelbrm — COMMENTED** [review]
Also, the branch has not been updated since the last merge to `master`, and is currently in conflict. This needs to be resolved too.

File: services/skus/service_nonint_test.go (hunk 4 above)
Comment (pavelbrm) [reply services/skus/service_nonint_test.go:4358]: We have discussed this, and after the conversation I see the benefits of using it in that particular instance.

File: services/skus/controllers_test.go [hunk 5]
```diff
@@ -1759,6 +1760,72 @@ func (suite *ControllersTestSuite) TestCreateOrderCreds_SingleUse_ExistingOrderC
 	suite.Assert().Contains(appError.Error(), ErrCredsAlreadyExist.Error())
 }
 
+func (suite *ControllersTestSuite) TestCreateOrder_RadomPayable() {
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/controllers_test.go:1763]: Discussed this, and it makes since it's an integration test. Resolving.

File: services/skus/controllers.go [hunk 6]
```diff
@@ -19,15 +18,13 @@ import (
...
 
-	"github.com/brave-intl/bat-go/libs/clients/radom"
 	appctx "github.com/brave-intl/bat-go/libs/context"
 	"github.com/brave-intl/bat-go/libs/handlers"
 	"github.com/brave-intl/bat-go/libs/inputs"
 	"github.com/brave-intl/bat-go/libs/logging"
 	"github.com/brave-intl/bat-go/libs/middleware"
 	"github.com/brave-intl/bat-go/libs/requestutils"
 	"github.com/brave-intl/bat-go/libs/responses"
-
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/controllers.go:30]: IDE has gone mad, will fix the imports

File: services/skus/model/model_pvt_test.go [hunk 7]
```diff
@@ -26,7 +26,7 @@ func TestAddURLParam(t *testing.T) {
 		exp   tcExpected
 	}
 
-	// Don't test for invalid inputs due to url.Parse's tolerance.
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/model/model_pvt_test.go:29]: New IDE is correcting spelling...

File: services/skus/service.go [hunk 8]
```diff
@@ -1976,14 +1977,72 @@ func (s *Service) createStripeSessID(ctx context.Context, req *model.CreateOrder
...
+		Gateway:    s.radomGateway,
+		SuccessURL: surl,
+		CancelURL:  curl,
+		Metadata: []radom.Metadata{
+			{
+				Key:   "brave_order_id",
+				Value: oid,
+			},
+		},
+		ExpiresAt: time.Now().Add(24 * time.Hour).Unix(),
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/service.go:1967]: The 24 hrs is added keep the checkout expiry consistent with the other payment providers. Will add UTC. I don't think the extra indirection is necessary atm moment as it just gets passed in and we don't need to assert on it in the tests.

File: services/skus/service.go [hunk 9]
```diff
@@ -1976,14 +1977,72 @@ func (s *Service) createStripeSessID(ctx context.Context, req *model.CreateOrder
...
+
+	return resp.SessionID, nil
+}
+
+const (
+	errRadomProductIDNotFound = model.Error("product id not found in metadata")
+	errRadomInvalidType       = model.Error("invalid type for product id")
+)
+
+func orderItemsToLineItems(orderItems []model.OrderItem) ([]radom.LineItem, error) {
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/service.go:1983]: Agreed, the name is a little generic although imo this function is mapping i.e. converting a domain object to a Data Transfer Object (DTO) opposed to 'building' which implies some sort of complexity to its creation e.g business logic, multiple object etc hence the `To`. Alternatively, `orderItemsToRadomLineItems` or `radomLineItemsFromOrderItems` might make more sense given we are just mapping To/From.

File: services/skus/radom/radom.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+type CheckoutSessionRequest struct {
+	LineItems  []LineItem `json:"lineItems"`
+	Gateway    Gateway    `json:"gateway"`
+	SuccessURL string     `json:"successUrl"`
+	CancelURL  string     `json:"cancelUrl"`
+	Metadata   []Metadata `json:"metadata"`
+	ExpiresAt  int64      `json:"expiresAt"` // in unix seconds
+}
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/radom/radom.go:53]: The fields are ordered by importance and the proposed order places the least important fields at the top which seems counter intuitive even if it is subjectively more pleasing to the eye. If anything we should consider field alignment. Going to leave this for now and maybe we can consider using field alignment going forward.

File: services/skus/service.go [hunk 11]
```diff
@@ -1976,14 +1977,72 @@ func (s *Service) createStripeSessID(ctx context.Context, req *model.CreateOrder
...
+	for i := range orderItems {
+		m, ok := orderItems[i].Metadata["radom_product_id"]
+		if !ok {
+			return nil, errRadomProductIDNotFound
+		}
+
+		pid, ok := m.(string)
+		if !ok {
+			return nil, errRadomInvalidType
+		}
```
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/service.go:1994]: This function was simply mapping from one representation to the other and doing some simple validation along the way. Given the `Metadata` is a `map[string]interface{}` it is entirely possible to store the incorrect type hence the specific type check. I can see the benefit of having a 'getter' for the product_id so will update to keep consistent with the stripe implementation.

File: services/skus/service.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/service.go]: Is the call to UTC necessary given the `.Unix()` returns the seconds elapsed since Jan 1970 UTC. 
> The result does not depend on the location associated with t.

File: services/skus/service.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/service.go]: Agreed, let's go with `orderItemsToRadomLineItems`. And we can later update the corresponding Stripe function to `orderItemsToStripeLineItems`.

File: services/skus/radom/radom.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+type CheckoutSessionResponse struct {
+	SessionID  string `json:"checkoutSessionId"`
+	SessionURL string `json:"checkoutSessionUrl"`
+}
+
+func (c *Client) CreateCheckoutSession(ctx context.Context, creq CheckoutSessionRequest) (CheckoutSessionResponse, error) {
```
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/radom/radom.go]: The request is now passed by pointer, but the original comment suggested returning the result by pointer as well. What do you think about it (asking because that has not changed)?

File: services/skus/service.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pavelbrm) [reply services/skus/service.go]: Noticed that `orderItemsToRadomLineItems` is already used, resolving.

File: services/skus/radom/radom.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (clD11) [reply services/skus/radom/radom.go]: The `CheckoutResponse` has been implemented to behave like a value object so I was going to leave that.

**Review by pavelbrm — APPROVED** [review]
Good work, approved! 👍
